File: ambeth/Ambeth.Xml/ambeth/xml/CyclicXmlController.cs
using System;
using System.Collections.Generic;
using De.Osthus.Ambeth.Ioc.Extendable;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Merge;
using De.Osthus.Ambeth.Util;
using De.Osthus.Ambeth.Xml.Typehandler;

namespace De.Osthus.Ambeth.Xml
{
    public class CyclicXmlController : AbstractHandler, ICyclicXmlController, ITypeBasedHandlerExtendable, INameBasedHandlerExtendable
    {
        [LogInstance]
        public new ILogger Log { private get; set; }

        public virtual IProxyHelper ProxyHelper { protected get; set; }

        public virtual IXmlTypeRegistry XmlTypeRegistry { protected get; set; }

        protected IMapExtendableContainer<Type, ITypeBasedHandler> typeToElementHandlers = new ClassExtendableContainer<ITypeBasedHandler>("elementHandler", "type");

        protected IMapExtendableContainer<String, INameBasedHandler> nameBasedElementReaders = new MapExtendableContainer<String, INameBasedHandler>("nameBasedElementReader", "elementName");

        protected IList<INameBasedHandler> nameBasedElementWriters = new List<INameBasedHandler>();

        public override void AfterPropertiesSet()
        {
            base.AfterPropertiesSet();

            ParamChecker.AssertNotNull(ProxyHelper, "ProxyHelper");
            ParamChecker.AssertNotNull(XmlTypeRegistry, "XmlTypeRegistry");
        }

        public virtual void RegisterElementHandler(ITypeBasedHandler elementHandler, Type type)
        {
            typeToElementHandlers.Register(elementHandler, type);
        }

        public virtual void UnregisterElementHandler(ITypeBasedHandler elementHandler, Type type)
        {
            typeToElementHandlers.Unregister(elementHandler, type);
        }

        public virtual void RegisterNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
        {
            nameBasedElementReaders.Register(nameBasedElementHandler, elementName);
            nameBasedElementWriters.Add(nameBasedElementHandler);
        }

        public virtual void UnregisterNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
        {
            nameBasedElementReaders.Unregister(nameBasedElementHandler, elementName);
            nameBasedElementWriters.Remove(nameBasedElementHandler);
        }

        public virtual Object ReadObject(IReader reader)
        {
            return ReadObject(typeof(Object), reader);
        }

        public virtual Object ReadObject(Type returnType, IReader reader)
        {
            String elementName = reader.GetElementName();
            if (XmlDictionary.NullElement.Equals(elementName))
            {
                reader.MoveOverElementEnd();
                return null;
            }
            String idValue = reader.GetAttributeValue(XmlDictionary.IdAttribute);
            int id = idValue != null && idValue.Length > 0 ? Int32.Parse(idValue) : 0;
            if (XmlDictionary.RefElement.Equals(elementName))
            {
                reader.MoveOverElementEnd();
                return reader.GetObjectById(id);
            }
            Object obj;
            if (XmlDictionary.ObjectElement.Equals(elementName))
            {
                Type type = ClassElementHandler.ReadFromAttribute(reader);
                obj = ReadObjectContent(returnType, type, id, reader);
            }
            else
            {
                INameBasedHandler nameBasedElementReader = nameBasedElementReaders.GetExtension(elementName);
                if (nameBasedElementReader == null)
                {
                    throw new Exception("Element name '" + elementName + "' not supported");
                }
                obj = nameBasedElementReader.ReadObject(returnType, elementName, id, reader);
            }
            if (id > 0)
            {
                reader.PutObjectWithId(obj, id);
            }
            reader.MoveOverElementEnd();
            return obj;
        }

        public virtual void WriteObject(Object obj, IWriter writer)
        {
            if (obj == null)
            {
                writer.WriteStartElement(XmlDictionary.NullElement);
                writer.WriteEndElement();
                return;
            }
            int id = writer.GetIdOfObject(obj);
            if (id > 0)
            {
                writer.WriteStartElement(XmlDictionary.RefElement);
                writer.WriteAttribute(XmlDictionary.IdAttribute, id.ToString());
                writer.WriteEndElement();
                return;
            }
            Type type = ProxyHelper.GetRealType(obj.GetType());
            for (int a = 0, size = nameBasedElementWriters.Count; a < size; a++)
            {
                INameBasedHandler nameBasedElementWriter = nameBasedElementWriters[a];
                if (nameBasedElementWriter.WritesCustom(obj, type, writer))
                {
                    return;
                }
            }

            id = writer.AcquireIdForObject(obj);
            String objectElement = XmlDictionary.ObjectElement;
            writer.WriteStartElement(objectElement);
            if (id > 0)
            {
                writer.WriteAttribute(XmlDictionary.IdAttribute, id.ToString());
            }
            ClassElementHandler.WriteAsAttribute(type, writer);
            WriteObjectContent(obj, type, writer);
            writer.WriteCloseElement(objectElement);
        }

        protected virtual Object ReadObjectContent(Type returnType, Type type, int id, IReader reader)
        {
            ITypeBasedHandler extension = typeToElementHandlers.GetExtension(type);
            if (extension == null)
            {
                throw new Exception("No extension mapped to type '" + type.Name + "' found");
            }
            return extension.ReadObject(returnType, type, id, reader);
        }

        protected virtual void WriteObjectContent(Object obj, Type type, IWriter writer)
        {
            ITypeBasedHandler extension = typeToElementHandlers.GetExtension(type);
            if (extension == null)
            {
                throw new Exception("No extension mapped to type '" + type.Name + "' found");
            }
            extension.WriteObject(obj, type, writer);
        }

        protected virtual Object ResolveObjectById(IDictionary<int, Object> idToObjectMap, int id)
        {
            Object obj = DictionaryExtension.ValueOrDefault(idToObjectMap, id);
            if (obj == null)
            {
                throw new Exception("No object found with id " + id);
            }
            return obj;
        }
    }
}

File: jambeth/jambeth-xml/src/main/java/de/osthus/ambeth/xml/CyclicXmlController.java
package de.osthus.ambeth.xml;

import java.util.List;

import de.osthus.ambeth.collections.ArrayList;
import de.osthus.ambeth.collections.IntKeyMap;
import de.osthus.ambeth.ioc.extendable.ClassExtendableContainer;
import de.osthus.ambeth.ioc.extendable.MapExtendableContainer;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.merge.IProxyHelper;
import de.osthus.ambeth.util.ParamChecker;
import de.osthus.ambeth.xml.typehandler.AbstractHandler;

public class CyclicXmlController extends AbstractHandler implements ICyclicXmlController, ITypeBasedHandlerExtendable, INameBasedHandlerExtendable
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	protected IProxyHelper proxyHelper;

	protected IXmlTypeRegistry xmlTypeRegistry;

	protected final ClassExtendableContainer<ITypeBasedHandler> typeToElementHandlers = new ClassExtendableContainer<ITypeBasedHandler>("elementHandler",
			"type");

	protected final MapExtendableContainer<String, INameBasedHandler> nameBasedElementReaders = new MapExtendableContainer<String, INameBasedHandler>(
			"nameBasedElementReader", "elementName");

	protected final List<INameBasedHandler> nameBasedElementWriters = new ArrayList<INameBasedHandler>();

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		super.afterPropertiesSet();

		ParamChecker.assertNotNull(proxyHelper, "proxyHelper");
		ParamChecker.assertNotNull(xmlTypeRegistry, "xmlTypeRegistry");
	}

	public void setProxyHelper(IProxyHelper proxyHelper)
	{
		this.proxyHelper = proxyHelper;
	}

	public void setXmlTypeRegistry(IXmlTypeRegistry xmlTypeRegistry)
	{
		this.xmlTypeRegistry = xmlTypeRegistry;
	}

	@Override
	public void registerElementHandler(ITypeBasedHandler elementHandler, Class<?> type)
	{
		typeToElementHandlers.register(elementHandler, type);
	}

	@Override
	public void unregisterElementHandler(ITypeBasedHandler elementHandler, Class<?> type)
	{
		typeToElementHandlers.unregister(elementHandler, type);
	}

	@Override
	public void registerNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
	{
		nameBasedElementReaders.register(nameBasedElementHandler, elementName);
		nameBasedElementWriters.add(nameBasedElementHandler);
	}

	@Override
	public void unregisterNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
	{
		nameBasedElementReaders.unregister(nameBasedElementHandler, elementName);
		nameBasedElementWriters.remove(nameBasedElementHandler);
	}

	@Override
	public Object readObject(IReader reader)
	{
		return readObject(Object.class, reader);
	}

	@Override
	public Object readObject(Class<?> returnType, IReader reader)
	{
		String elementName = reader.getElementName();
		if (xmlDictionary.getNullElement().equals(elementName))
		{
			reader.moveOverElementEnd();
			return null;
		}
		String idValue = reader.getAttributeValue(xmlDictionary.getIdAttribute());
		int id = idValue != null && idValue.length() > 0 ? Integer.parseInt(idValue) : 0;
		if (xmlDictionary.getRefElement().equals(elementName))
		{
			reader.moveOverElementEnd();
			return reader.getObjectById(id);
		}
		Object obj;
		if (xmlDictionary.getObjectElement().equals(elementName))
		{
			Class<?> type = classElementHandler.readFromAttribute(reader);
			obj = readObjectContent(returnType, type, id, reader);
		}
		else
		{
			INameBasedHandler nameBasedElementReader = nameBasedElementReaders.getExtension(elementName);
			if (nameBasedElementReader == null)
			{
				throw new IllegalStateException("Element name '" + elementName + "' not supported");
			}
			obj = nameBasedElementReader.readObject(returnType, elementName, id, reader);
		}
		if (id > 0)
		{
			reader.putObjectWithId(obj, id);
		}
		reader.moveOverElementEnd();
		return obj;
	}

	@Override
	public void writeObject(Object obj, IWriter writer)
	{
		if (obj == null)
		{
			writer.writeStartElement(xmlDictionary.getNullElement());
			writer.writeEndElement();
			return;
		}
		Integer idValue = writer.getIdOfObject(obj);
		if (idValue > 0)
		{
			writer.writeStartElement(xmlDictionary.getRefElement());
			writer.writeAttribute(xmlDictionary.getIdAttribute(), idValue.toString());
			writer.writeEndElement();
			return;
		}
		Class<?> type = proxyHelper.getRealType(obj.getClass());
		for (int a = 0, size = nameBasedElementWriters.size(); a < size; a++)
		{
			INameBasedHandler nameBasedElementWriter = nameBasedElementWriters.get(a);
			if (nameBasedElementWriter.writesCustom(obj, type, writer))
			{
				return;
			}
		}
		int id = writer.acquireIdForObject(obj);
		String objectElement = xmlDictionary.getObjectElement();
		writer.writeStartElement(objectElement);
		if (id > 0)
		{
			writer.writeAttribute(xmlDictionary.getIdAttribute(), Integer.toString(id));
		}
		classElementHandler.writeAsAttribute(type, writer);
		writeObjectContent(obj, type, writer);
		writer.writeCloseElement(objectElement);
	}

	protected Object readObjectContent(Class<?> returnType, Class<?> type, int id, IReader reader)
	{
		ITypeBasedHandler extension = typeToElementHandlers.getExtension(type);
		if (extension == null)
		{
			throw new IllegalStateException("No extension mapped to type '" + type.getName() + "' found");
		}
		return extension.readObject(returnType, type, id, reader);
	}

	protected void writeObjectContent(Object obj, Class<?> type, IWriter writer)
	{
		ITypeBasedHandler extension = typeToElementHandlers.getExtension(type);
		if (extension == null)
		{
			throw new IllegalStateException("No extension mapped to type '" + type.getName() + "' found");
		}
		extension.writeObject(obj, type, writer);
	}

	protected Object resolveObjectById(IntKeyMap<Object> idToObjectMap, int id)
	{
		Object obj = idToObjectMap.get(id);
		if (obj == null)
		{
			throw new IllegalStateException("No object found with id " + id);
		}
		return obj;
	}
}